### PR TITLE
Implemented state transition according to Spec 0.7.1

### PR DESCRIPTION
**Pull request by sadiq1971**
* modified / removed functions according to spec 0.7.1
* modified data structure Crosslink and AttestationData according to spec.
````
export interface Crosslink {
  //Shard number
  shard: number64;
  //Crosslinking data from epochs [start....end-1]
  startEpoch: number64;
  endEpoch: number64;
  //Root of the previous crosslink
  parentRoot: Bytes32;
  //Root of the crosslinked shard data since the previous crosslink
  dataRoot: Bytes32;
}
````
In most of the cases i assumed startEpoc: GENESIS_EPOC. not sure about endEpoch.
* added spec code for better visibility while reviewing. I will remove those comment once review is done
* Removed some unit tests as some function is gone.
* Modified the tests also

### CODE REVIEW

**Comment by GregTheGreek** [comment]
Can we change the merge branch target to `dev/0.7.1`.

Rational:
Until **ALL** the 0.7.1 changes have been made I want to avoid polluting master with a failing build. So we can merge the subsequent updates into `dev/0.7.1` and then also fix the tests as we go as well.

File: src/chain/factory/block/index.ts
Comment (mpetrunic) [inline src/chain/factory/block/index.ts]: leave verify false flag here. It will fail otherwise because stateRoot and signatures are empty

File: src/chain/stateTransition/block/attestations.ts
Comment (mpetrunic) [inline src/chain/stateTransition/block/attestations.ts]: buffers should be compared with equals

File: src/chain/stateTransition/block/index.ts
Comment (mpetrunic) [inline src/chain/stateTransition/block/index.ts]: leave verify flag

File: src/chain/stateTransition/epoch/util.ts
Comment (mpetrunic) [inline src/chain/stateTransition/epoch/util.ts]: const output = attestations.map((a) => a.data.crosslink)

It's faster usually

File: src/chain/stateTransition/epoch/util.ts
Comment (mpetrunic) [inline src/chain/stateTransition/epoch/util.ts]: You cannot compare objects like that. For now serialize it using ssz and comapre buffers. When released, new ssz will have method equals for those objects

File: src/chain/stateTransition/epoch/util.ts
Comment (mpetrunic) [inline src/chain/stateTransition/epoch/util.ts]: canot compare like that

File: src/chain/stateTransition/index.ts
Comment (mpetrunic) [inline src/chain/stateTransition/index.ts:30]: Check tests, in current spec some of them are passing null blocks to advance slots

File: src/chain/stateTransition/slot.ts
Comment (mpetrunic) [inline src/chain/stateTransition/slot.ts]: use equals

File: src/chain/stateTransition/slot.ts
Comment (mpetrunic) [inline src/chain/stateTransition/slot.ts]: ===

File: src/chain/stateTransition/index.ts
Comment (mpetrunic) [inline src/chain/stateTransition/index.ts]: use equals method

File: src/types/misc.ts
Comment (mpetrunic) [inline src/types/misc.ts]: Bytes32 should be lowercase

**Review by mpetrunic — CHANGES_REQUESTED** [review]
You should change spec test submodule url and commit hash to official ethereum repo and 0.7.1 commit hash

File: src/chain/factory/block/index.ts
Comment (sadiq1971) [reply src/chain/factory/block/index.ts]: changed submodule url and updated processBlock with verify flag

File: src/chain/stateTransition/index.ts
Comment (wemeetagain) [reply src/chain/stateTransition/index.ts]: We'll need to update `@chainsafe/ssz` to v0.4.2

Comment (GregTheGreek) [reply src/chain/stateTransition/index.ts]: Make an issue :)

Comment (GregTheGreek) [reply src/chain/stateTransition/index.ts]: oih nvm thats already deployed

File: .gitmodules
Comment (GregTheGreek) [inline .gitmodules]: @mpetrunic @wemeetagain AFAIK this should remain chainsafe, perhaps for the purposes of this branch `dev/0.7.1` we can change this line to be a branch with the updated tests?

File: src/chain/stateTransition/index.ts
Comment (GregTheGreek) [reply src/chain/stateTransition/index.ts:30]: should we not also play catchup with the while loop from `6.x` code?

**Review by GregTheGreek — DISMISSED** [review]
~I think we should be good to merge this in (after my comments) because everything else is failing since its not updated to spec 0.7~

I lied, it looks as though there are still a substantial amount of tests failing WRT forkchoice

File: .gitmodules
Comment (mpetrunic) [reply .gitmodules]: yup, I told him to reference ethereum repo for simplicity until it's ready to merge

File: src/chain/stateTransition/index.ts
Comment (sadiq1971) [reply src/chain/stateTransition/index.ts:30]: The changes that have been done according to 0.7.1 spec is-
* parameter's defult value `validateStateRoot = false`  which was `true` before.
* added funtion `processSlots` - where there is the same while loop and equivalent code of `cacheState` and `advanceSlot ` 
* The only difference is that there is no if-else condition for checking null blocks. Instead there is an assert - `assert(state.slot <= slot)`
* Now what change should i made?

Comment (mpetrunic) [reply src/chain/stateTransition/index.ts:30]: That's fine, you should change spec test, where sending empty block was used to advance slot, to call processSlots method instead of stateTransition

File: src/chain/stateTransition/block/attestations.ts
Comment (sadiq1971) [inline src/chain/stateTransition/block/attestations.ts]: If i console log ffgData with right side value it's always same. But the comparing always gives false. what's the right way of comparing then?

File: src/chain/stateTransition/block/attestations.ts
Comment (sadiq1971) [inline src/chain/stateTransition/block/attestations.ts]: If i console log ffgData with right side value it's always same. But the comparing always gives false. what's the right way of comparing then?

Comment (mpetrunic) [reply src/chain/stateTransition/block/attestations.ts]: it compares it by reference. Trying JSON.stringify it

Comment (mpetrunic) [reply src/chain/stateTransition/block/attestations.ts]: maybe better is to `serialize(ffgData, [number64])`

Comment (wemeetagain) [reply src/chain/stateTransition/block/attestations.ts]: mark as TODO to use `ssz.equals`. It currently has a bug that effects this case, but we can use it soonish.

Comment (mpetrunic) [reply src/chain/stateTransition/block/attestations.ts]: I already have a branch where I replaced everything with equals :)

File: .gitmodules
Comment (GregTheGreek) [reply .gitmodules]: So should tests be passing ?

File: src/chain/stateTransition/block/attestations.ts
Comment (sadiq1971) [reply src/chain/stateTransition/block/attestations.ts]: > ffgData

but `ffgData` is array of `[number64, buffer, number64]`. So, should it be `serialize(ffgData, [number64])` ?

File: .gitmodules
Comment (sadiq1971) [reply .gitmodules]: I am updating spec tests. There are some errors. Trying to remove them.

File: src/chain/stateTransition/block/attestations.ts
Comment (mpetrunic) [reply src/chain/stateTransition/block/attestations.ts]: oh, true, hmm, thats annoying, I don't think we can serialize array of different types, @wemeetagain ?

Comment (wemeetagain) [reply src/chain/stateTransition/block/attestations.ts]: No, we can either compare each array index separately, or turn the arrays into objects.

File: .gitmodules
Comment (sadiq1971) [reply .gitmodules]: Data structures have been changed in 0.7.1. but I only updated it for stateTransition. So, spec tests are failing. I think if we have a complete merged version  then all the spec tests will pass.

File: src/chain/stateTransition/block/attestations.ts
Comment (mpetrunic) [reply src/chain/stateTransition/block/attestations.ts]: ffgData would be a nice object type though

File: .gitmodules
Comment (mpetrunic) [inline .gitmodules]: Shell we update chainsafe repo. I like greg idea with creating 0.7.1 branch in spec test repo

File: src/chain/factory/block/deposits.ts
Comment (mpetrunic) [inline src/chain/factory/block/deposits.ts]: you should add `+state.depositIndex`

File: src/chain/chain.ts
Comment (mpetrunic) [inline src/chain/chain.ts]: you should add `+newState.depositIndex`

File: src/chain/stateTransition/block/operations/attestation.ts
Comment (mpetrunic) [inline src/chain/stateTransition/block/operations/attestation.ts]: remove?

File: src/chain/stateTransition/util/attestation.ts
Comment (mpetrunic) [inline src/chain/stateTransition/util/attestation.ts]: I would move it to utils

File: test/unit/chain/factory/attestation/data.test.ts
Comment (mpetrunic) [inline test/unit/chain/factory/attestation/data.test.ts]: ?

File: test/unit/chain/factory/attestation/index.test.ts
Comment (mpetrunic) [inline test/unit/chain/factory/attestation/index.test.ts]: attestation production doesn't work or is changed?

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Hmm I saw that you used ssz.equals on some places. I guess you have linked ssz locally.
Every spec test result can use equals as comparison, and there are some other places in spec.

Maybe try to merge this branch https://github.com/ChainSafe/lodestar/tree/mpetrunic/optimize-object-comparison ?

File: test/unit/chain/factory/attestation/index.test.ts
Comment (wemeetagain) [reply test/unit/chain/factory/attestation/index.test.ts]: didn't catch this, will uncomment and see what happens

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**
